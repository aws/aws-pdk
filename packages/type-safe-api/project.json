{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
