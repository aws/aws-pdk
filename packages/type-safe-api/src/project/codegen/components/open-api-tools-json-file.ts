/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { ProjectUtils } from "@aws-prototyping-sdk/nx-monorepo";
import { JsonFile, Project } from "projen";
import { OpenApiGeneratorCliConfig } from "../../types";

/**
 * Represents an openapitools.json file
 * @see https://github.com/OpenAPITools/openapi-generator-cli#configuration
 */
export class OpenApiToolsJsonFile extends JsonFile {
  /**
   * Retrieves an instance of OpenApiToolsJsonFile if one is associated to the given project.
   *
   * @param project project instance.
   */
  static of(project: Project): OpenApiToolsJsonFile | undefined {
    return project.components.find((c) =>
      ProjectUtils.isNamedInstanceOf(c, OpenApiToolsJsonFile)
    ) as OpenApiToolsJsonFile | undefined;
  }

  /**
   * Retrieves an instance of OpenApiToolsJsonFile if one is associated to the given project,
   * otherwise creates a OpenApiToolsJsonFile instance for the project.
   *
   * @param project project instance.
   */
  static ensure(project: Project): OpenApiToolsJsonFile {
    return (
      OpenApiToolsJsonFile.of(project) || new OpenApiToolsJsonFile(project)
    );
  }

  /**
   * Configuration for OpenAPI Generator CLI
   * @private
   */
  private config: OpenApiGeneratorCliConfig = {
    version: "6.3.0",
    storageDir: "~/.open-api-generator-cli",
  };

  constructor(project: Project) {
    if (OpenApiToolsJsonFile.of(project)) {
      throw new Error(
        `Project ${project.name} already has associated OpenApiToolsJsonFile component.`
      );
    }

    super(project, "openapitools.json", {
      obj: {
        // Schema is located in node_modules when generator cli is installed in tmp dir
        $schema:
          "node_modules/@openapitools/openapi-generator-cli/config.schema.json",
        spaces: 2,
        "generator-cli": () => this.config,
      },
    });
  }

  /**
   * Update the OpenAPI Generator config. Shallow-merges with any existing default config
   */
  public addOpenApiGeneratorCliConfig(config?: OpenApiGeneratorCliConfig) {
    this.config = {
      ...this.config,
      ...config,
    };
  }
}
