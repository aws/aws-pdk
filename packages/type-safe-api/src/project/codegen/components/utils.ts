/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import * as readPkg from "read-pkg-up";
import { Language, Library } from "../../languages";
import { MockResponseDataGenerationOptions } from "../../types";

/**
 * Enum for generator directories for non-runtime generators
 */
export enum OtherGenerators {
  DOCS = "docs",
  // Infrastructure
  TYPESCRIPT_CDK_INFRASTRUCTURE = "typescript-cdk-infrastructure",
  PYTHON_CDK_INFRASTRUCTURE = "python-cdk-infrastructure",
  JAVA_CDK_INFRASTRUCTURE = "java-cdk-infrastructure",
  // Handlers
  TYPESCRIPT_LAMBDA_HANDLERS = "typescript-lambda-handlers",
  PYTHON_LAMBDA_HANDLERS = "python-lambda-handlers",
  JAVA_LAMBDA_HANDLERS = "java-lambda-handlers",
}

export enum TypeSafeApiScript {
  PARSE_OPENAPI_SPEC = "parse-openapi-spec",
  GENERATE = "generate",
  GENERATE_MOCK_DATA = "generate-mock-data",
  GENERATE_HTML_REDOC_DOCS = "generate-html-redoc-docs",
  CLEAN_OPENAPI_GENERATED_CODE = "clean-openapi-generated-code",
  COPY_GRADLE_WRAPPER = "copy-gradle-wrapper",
}

/**
 * Generator directory for openapi generation containing templates, config etc.
 */
export type GeneratorDirectory = Language | Library | OtherGenerators;

/**
 * Types of normalizers supported by openapi-generator
 * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
 */
export type OpenApiNormalizer = "KEEP_ONLY_FIRST_TAG_IN_OPERATION";

/**
 * Options for generating code or docs using OpenAPI Generator CLI
 */
export interface GenerationOptions {
  /**
   * The OpenAPI generator to use to generate the code/docs
   */
  readonly generator: string;
  /**
   * The directory to use for OpenAPI generation
   */
  readonly generatorDirectory: GeneratorDirectory;
  /**
   * The path of the OpenAPI spec to generate code from
   */
  readonly specPath: string;
  /**
   * Additional properties to pass to the generate cli
   */
  readonly additionalProperties?: {
    [key: string]: string;
  };
  /**
   * Supply the relative path from the code project root to the source code directory in which custom generated files
   * (eg. operation config) should be placed.
   */
  readonly srcDir?: string;
  /**
   * Normalizers to apply to the spec prior to generation, if any
   * @see https://openapi-generator.tech/docs/customization/#openapi-normalizer
   */
  readonly normalizers?: Partial<Record<OpenApiNormalizer, boolean>>;
  /**
   * Vendor extensions to add for code generation, allowing custom properties to be passed to the generator templates
   * Keys should begin with "x-"
   */
  readonly extraVendorExtensions?: Record<string, string | boolean>;
  /**
   * Generate alias as model.
   * Defines whether "primitive types" defined at the model/schema level will be wrapped in a model (ie maps/lists)
   * @see https://openapi-generator.tech/docs/globals/
   * @default true
   */
  readonly generateAliasAsModel?: boolean;
  /**
   * The path to the json smithy model file, if available
   */
  readonly smithyJsonPath?: string;
}

/**
 * Build a command for running a script from this project's bin
 */
export const buildTypeSafeApiExecCommand = (
  script: TypeSafeApiScript,
  args?: string
) => {
  const { packageJson } = readPkg.sync({
    cwd: path.resolve(__dirname),
  })!;
  return `npx --yes -p aws-pdk@${packageJson.version} ${script}${
    args ? ` ${args}` : ""
  }`;
};

const serializeProperties = (properties: { [key: string]: string }) =>
  Object.entries(properties)
    .map(([key, value]) => `${key}=${value}`)
    .join(",");

/**
 * Generate code or docs by invoking the root generate script
 */
export const buildInvokeOpenApiGeneratorCommandArgs = (
  options: GenerationOptions
): string => {
  const srcDir = options.srcDir ?? "src";
  const additionalProperties = options.additionalProperties
    ? ` --additional-properties "${serializeProperties(
        options.additionalProperties
      )}"`
    : "";

  const normalizers = options.normalizers
    ? ` --openapi-normalizer "${serializeProperties(
        Object.fromEntries(
          Object.entries(options.normalizers).map(([k, v]) => [k, `${v}`])
        )
      )}"`
    : "";

  const extensions = options.extraVendorExtensions
    ? ` --extra-vendor-extensions '${JSON.stringify(
        options.extraVendorExtensions
      )}'`
    : "";

  const generateAliasAsModel =
    options.generateAliasAsModel ?? true ? " --generate-alias-as-model" : "";

  const smithyJsonPath = options.smithyJsonPath
    ? ` --smithy-json-path ${options.smithyJsonPath}`
    : "";

  const specPath = options.specPath;
  const outputPath = ".";

  return `--generator ${options.generator} --spec-path ${specPath} --output-path ${outputPath} --generator-dir ${options.generatorDirectory} --src-dir ${srcDir}${smithyJsonPath}${additionalProperties}${normalizers}${extensions}${generateAliasAsModel}`;
};

/**
 * Builds a command to clean up files which were previously generated by openapi generator
 */
export const buildCleanOpenApiGeneratedCodeCommand = (): string => {
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.CLEAN_OPENAPI_GENERATED_CODE,
    `--code-path .`
  );
};

/**
 * Options for generating mock data json files
 */
export interface MockDataGenerationOptions
  extends MockResponseDataGenerationOptions {
  /**
   * The path of the OpenAPI spec to generate data for
   */
  readonly specPath: string;
  /**
   * Output sub directory relative to the outdir in which to generate mock data
   * Mock data will be written to a directory named 'mocks' within the sub directory
   * @default .
   */
  readonly outputSubDir?: string;
}

/**
 * Invoke the mock data generator script
 */
export const buildInvokeMockDataGeneratorCommand = (
  options: MockDataGenerationOptions
): string => {
  const outputPath = options.outputSubDir ?? ".";
  const locale = options.locale ? ` --locale ${options.locale}` : "";
  const maxArrayLength =
    options.maxArrayLength !== undefined
      ? ` --max-array-length ${options.maxArrayLength}`
      : "";
  return buildTypeSafeApiExecCommand(
    TypeSafeApiScript.GENERATE_MOCK_DATA,
    `--spec-path ${options.specPath} --output-path ${outputPath}${locale}${maxArrayLength}`
  );
};
