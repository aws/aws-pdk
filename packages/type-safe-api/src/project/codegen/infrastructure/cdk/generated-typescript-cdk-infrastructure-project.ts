/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import {
  GeneratedTypescriptCdkInfrastructureBaseProject,
  GeneratedTypescriptCdkInfrastructureBaseProjectOptions,
} from "./generated-typescript-cdk-infrastructure-base-project";
import { MockResponseGenerationOptions } from "../../../types";
import {
  buildInvokeMockDataGeneratorCommand,
  CodegenOptions,
  getHandlersProjectVendorExtensions,
  OtherGenerators,
} from "../../components/utils";

export interface GeneratedTypescriptCdkInfrastructureProjectOptions
  extends GeneratedTypescriptCdkInfrastructureBaseProjectOptions,
    MockResponseGenerationOptions {}

export class GeneratedTypescriptCdkInfrastructureProject extends GeneratedTypescriptCdkInfrastructureBaseProject {
  /**
   * Options configured for the project
   * @private
   */
  protected readonly options: GeneratedTypescriptCdkInfrastructureProjectOptions;

  constructor(options: GeneratedTypescriptCdkInfrastructureProjectOptions) {
    super(options);
    this.options = options;

    if (!this.options.mockDataOptions?.disable) {
      this.generateTask.exec(this.buildGenerateMockDataCommand());
    }
  }

  protected buildCodegenOptions(): CodegenOptions {
    return {
      specPath: this.options.specPath,
      templateDirs: [OtherGenerators.TYPESCRIPT_CDK_INFRASTRUCTURE],
      metadata: {
        srcDir: this.srcdir,
        runtimePackageName:
          this.options.generatedTypescriptTypes.package.packageName,
        // Spec path relative to the source directory
        relativeSpecPath: path.join("..", this.packagedSpecPath),
        // Enable mock integration generation by default
        enableMockIntegrations: !this.options.mockDataOptions?.disable,
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
    };
  }

  public buildGenerateMockDataCommand = () => {
    return buildInvokeMockDataGeneratorCommand({
      specPath: this.options.specPath,
      ...this.options.mockDataOptions,
    });
  };
}
