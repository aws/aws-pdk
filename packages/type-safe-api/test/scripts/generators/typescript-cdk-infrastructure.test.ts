/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { getTestHandlerProjects } from "./utils";
import { OpenApiToolsJsonFile } from "../../../src/project/codegen/components/open-api-tools-json-file";
import { GeneratedTypescriptCdkInfrastructureProject } from "../../../src/project/codegen/infrastructure/cdk/generated-typescript-cdk-infrastructure-project";
import { GeneratedTypescriptRuntimeProject } from "../../../src/project/codegen/runtime/generated-typescript-runtime-project";
import { withTmpDirSnapshot } from "../../project/snapshot-utils";

describe("Typescript Infrastructure Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(__dirname, `../../resources/specs/${spec}`);

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedTypescriptRuntimeProject({
        name: "test-client",
        defaultReleaseBranch: "main",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedTypescriptCdkInfrastructureProject({
        name: "test-infra",
        defaultReleaseBranch: "main",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedTypescriptTypes: client,
        generatedHandlers: {},
      });
      // Synth the openapitools.json since it's used by the generate command
      OpenApiToolsJsonFile.of(project)!.synthesize();
      exec(
        `${path.resolve(
          __dirname,
          "../../../scripts/generators/generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/src/api.ts"]).toMatchSnapshot();
    expect(snapshot["infra/src/mock-integrations.ts"]).toMatchSnapshot();
    expect(snapshot["infra/src/index.ts"]).toMatchSnapshot();
  });

  it("Generates With Mocks Disabled", () => {
    const specPath = path.resolve(
      __dirname,
      `../../resources/specs/single.yaml`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedTypescriptRuntimeProject({
        name: "test-client",
        defaultReleaseBranch: "main",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedTypescriptCdkInfrastructureProject({
        name: "test-infra",
        defaultReleaseBranch: "main",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedTypescriptTypes: client,
        mockDataOptions: {
          disable: true,
        },
        generatedHandlers: {},
      });
      // Synth the openapitools.json since it's used by the generate command
      OpenApiToolsJsonFile.of(project)!.synthesize();
      exec(
        `${path.resolve(
          __dirname,
          "../../../scripts/generators/generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/src/mock-integrations.ts"]).toMatchSnapshot();
  });

  it("Generates Functions", () => {
    const specPath = path.resolve(
      __dirname,
      `../../resources/specs/handlers.yaml`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const { runtimes, handlers } = getTestHandlerProjects(outdir);
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedTypescriptCdkInfrastructureProject({
        name: "test-infra",
        defaultReleaseBranch: "main",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedTypescriptTypes: runtimes.typescript,
        generatedHandlers: handlers,
      });
      project.synth();
      exec(
        `${path.resolve(
          __dirname,
          "../../../scripts/generators/generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/src/functions.ts"]).toMatchSnapshot();
  });
});
