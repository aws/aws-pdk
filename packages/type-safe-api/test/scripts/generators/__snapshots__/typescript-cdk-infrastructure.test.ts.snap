// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates Functions for handlers.yaml 1`] = `
"import { Construct } from "constructs";
import { Duration } from "aws-cdk-lib";
import { SnapStartFunction, SnapStartFunctionProps } from "@aws/pdk/type-safe-api";
import { Code, Function, Runtime, Tracing, FunctionProps } from "aws-cdk-lib/aws-lambda";
import * as path from "path";

/**
 * Options for the JavaOneFunction construct
 */
export interface JavaOneFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of JavaOne
 */
export class JavaOneFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaOneHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the JavaTwoFunction construct
 */
export interface JavaTwoFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of JavaTwo
 */
export class JavaTwoFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaTwoHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the PythonOneFunction construct
 */
export interface PythonOneFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of PythonOne
 */
export class PythonOneFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python_one.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the PythonTwoFunction construct
 */
export interface PythonTwoFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of PythonTwo
 */
export class PythonTwoFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python_two.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the TypescriptOneFunction construct
 */
export interface TypescriptOneFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypescriptOne
 */
export class TypescriptOneFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypescriptOneFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "typescript-one",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the TypescriptTwoFunction construct
 */
export interface TypescriptTwoFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypescriptTwo
 */
export class TypescriptTwoFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypescriptTwoFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "typescript-two",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

"
`;

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates Functions for inline-body.yaml 1`] = `
"import { Construct } from "constructs";
import { Duration } from "aws-cdk-lib";
import { SnapStartFunction, SnapStartFunctionProps } from "@aws/pdk/type-safe-api";
import { Code, Function, Runtime, Tracing, FunctionProps } from "aws-cdk-lib/aws-lambda";
import * as path from "path";

/**
 * Options for the JavaTestFunction construct
 */
export interface JavaTestFunctionProps extends Omit<SnapStartFunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the java implementation of JavaTest
 */
export class JavaTestFunction extends SnapStartFunction {
  constructor(scope: Construct, id: string, props?: JavaTestFunctionProps) {
    super(scope, id, {
      runtime: Runtime.JAVA_17,
      handler: "test.test-java-handlers.handlers.JavaTestHandler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the PythonTestFunction construct
 */
export interface PythonTestFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the python implementation of PythonTest
 */
export class PythonTestFunction extends Function {
  constructor(scope: Construct, id: string, props?: PythonTestFunctionProps) {
    super(scope, id, {
      runtime: Runtime.PYTHON_3_11,
      handler: "test_python_handlers.python_test.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../python-handlers/dist/lambda",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

/**
 * Options for the TypescriptTestFunction construct
 */
export interface TypescriptTestFunctionProps extends Omit<FunctionProps, 'code' | 'handler' | 'runtime'> {}

/**
 * Lambda function construct which points to the typescript implementation of TypescriptTest
 */
export class TypescriptTestFunction extends Function {
  constructor(scope: Construct, id: string, props?: TypescriptTestFunctionProps) {
    super(scope, id, {
      runtime: Runtime.NODEJS_18_X,
      handler: "index.handler",
      code: Code.fromAsset(path.resolve(__dirname, "..",
        "../typescript-handlers/dist/lambda",
        "typescript-test",
      )),
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(30),
      ...props,
    });
  }
}

"
`;

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates With Mocks Disabled 1`] = `
"import {
  ApiError,
  ApiErrorToJSON,
  MapResponse,
  MapResponseToJSON,
  MapResponseMapPropertyValue,
  MapResponseMapPropertyValueToJSON,
  TestRequest,
  TestRequestToJSON,
  TestResponse,
  TestResponseToJSON,
  TestResponseMessagesInner,
  TestResponseMessagesInnerToJSON,
} from "test-client";
import { Integrations, MockIntegration } from "@aws/pdk/type-safe-api";
import * as fs from "fs";
import * as path from "path";

/**
 * Type-safe mock integrations for API operations
 */
export class MockIntegrations {
  /**
   * Read a mock data file for the given operation
   */
  private static readMockDataFile(method: string, urlPath: string, statusCode: number): string {
    return fs.readFileSync(path.join(__dirname, "..", "mocks", \`\${method.toLowerCase()}\${urlPath.replace(/\\//g, "-")}-\${statusCode}.json\`), "utf-8");
  }

  // No mock integrations have been generated, since mock data generation is disabled.

}
"
`;

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates With single.yaml 1`] = `
"import { TypeSafeRestApi, TypeSafeRestApiProps, TypeSafeApiIntegration } from "@aws/pdk/type-safe-api";
import { Construct } from "constructs";
import { OperationLookup, OperationConfig } from "test-client";
import * as path from "path";

export type ApiIntegrations = OperationConfig<TypeSafeApiIntegration>;

export interface ApiProps extends Omit<TypeSafeRestApiProps, "specPath" | "operationLookup" | "integrations"> {
  readonly integrations: ApiIntegrations;
}

/**
 * Type-safe construct for the API Gateway resources defined by your model.
 * This construct is generated and should not be modified.
 */
export class Api extends TypeSafeRestApi {
  constructor(scope: Construct, id: string, props: ApiProps) {
    super(scope, id, {
      ...props,
      integrations: props.integrations as any,
      specPath: path.resolve(__dirname, "../assets/api.json"),
      operationLookup: OperationLookup as any,
    });
  }
}
"
`;

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates With single.yaml 2`] = `
"import {
  ApiError,
  ApiErrorToJSON,
  MapResponse,
  MapResponseToJSON,
  MapResponseMapPropertyValue,
  MapResponseMapPropertyValueToJSON,
  TestRequest,
  TestRequestToJSON,
  TestResponse,
  TestResponseToJSON,
  TestResponseMessagesInner,
  TestResponseMessagesInnerToJSON,
} from "test-client";
import { Integrations, MockIntegration } from "@aws/pdk/type-safe-api";
import * as fs from "fs";
import * as path from "path";

/**
 * Type-safe mock integrations for API operations
 */
export class MockIntegrations {
  /**
   * Read a mock data file for the given operation
   */
  private static readMockDataFile(method: string, urlPath: string, statusCode: number): string {
    return fs.readFileSync(path.join(__dirname, "..", "mocks", \`\${method.toLowerCase()}\${urlPath.replace(/\\//g, "-")}-\${statusCode}.json\`), "utf-8");
  }

  /**
   * Mock integration to return a 200 response from the anyRequestResponse operation
   */
  public static anyRequestResponse200(body: string): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body,
    });
  }

  /**
   * Mock integration to return a 204 response from the empty operation
   */
  public static empty204(): MockIntegration {
    return Integrations.mock({
      statusCode: 204,
    });
  }

  /**
   * Mock integration to return a 200 response from the mapResponse operation
   * Call this with no arguments to use a generated mock response
   */
  public static mapResponse200(body?: MapResponse): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body: body === undefined
        ? MockIntegrations.readMockDataFile("GET", "/map-response", 200)
        : JSON.stringify(MapResponseToJSON(body)),
    });
  }

  /**
   * Mock integration to return a 200 response from the mediaTypes operation
   */
  public static mediaTypes200(body: string): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body,
    });
  }

  /**
   * Mock integration to return a 200 response from the multipleContentTypes operation
   */
  public static multipleContentTypes200(body: string): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body,
    });
  }

  /**
   * Mock integration to return a 200 response from the operationOne operation
   * Call this with no arguments to use a generated mock response
   */
  public static operationOne200(body?: TestResponse): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body: body === undefined
        ? MockIntegrations.readMockDataFile("POST", "/path/{pathParam}", 200)
        : JSON.stringify(TestResponseToJSON(body)),
    });
  }

  /**
   * Mock integration to return a 400 response from the operationOne operation
   * Call this with no arguments to use a generated mock response
   */
  public static operationOne400(body?: ApiError): MockIntegration {
    return Integrations.mock({
      statusCode: 400,
      body: body === undefined
        ? MockIntegrations.readMockDataFile("POST", "/path/{pathParam}", 400)
        : JSON.stringify(ApiErrorToJSON(body)),
    });
  }

  /**
   * Mock integration to return a 200 response from the withoutOperationIdDelete operation
   * Call this with no arguments to use a generated mock response
   */
  public static withoutOperationIdDelete200(body?: TestResponse): MockIntegration {
    return Integrations.mock({
      statusCode: 200,
      body: body === undefined
        ? MockIntegrations.readMockDataFile("DELETE", "/without-operation-id", 200)
        : JSON.stringify(TestResponseToJSON(body)),
    });
  }


  /**
   * Mock all operations for which generated JSON data can be returned.
   * The first available response is used for each operation. In most cases this is the successful 200 response.
   */
  public static mockAll() {
    return {
      mapResponse: {
        integration: MockIntegrations.mapResponse200(),
      },
      operationOne: {
        integration: MockIntegrations.operationOne200(),
      },
      withoutOperationIdDelete: {
        integration: MockIntegrations.withoutOperationIdDelete200(),
      },
    };
  }

}
"
`;

exports[`Typescript Infrastructure Code Generation Script Unit Tests Generates With single.yaml 3`] = `
"export * from "./api";
export * from "./mock-integrations";
export * from "./functions";
"
`;
